File: examples/hello_params/hello_params_py/hello_params_py/hello_params.py
#!/usr/bin/env python3

'''
    A minimal ROS2 python node that allows to declare and modify params.

    Basic structure for ROS2 node publisher:
    - Initialise rclpy
    - Create a node
    - Declare the params
    - Keep it alive with spin
    - Shutdown

    RELIABLE and VOLATILE have been selected for the QOS_PROFILE to match with
    the ones defined in the service server.

    Author: Ioannis Selinis <selinis.g@gmail.com> 2025
'''


import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from rcl_interfaces.msg import ParameterDescriptor, FloatingPointRange, SetParametersResult


class HelloParams(Node):
    """The HelloParams class that inherits from rclpy.node.Node.
    Initialises with the params.
    """
    def __init__(self):
        super().__init__(node_name='hello_params')

        # declare_parameters to declare multiple ones
        # use declare_parameter to declare a single
        self.declare_parameters(namespace='',
                                parameters=[
                                    ('timer_interval', 1.0, ParameterDescriptor(description='Set the interval for the timer',
                                                                                floating_point_range=[FloatingPointRange(from_value=1.0,
                                                                                                                         to_value=13.0,
                                                                                                                         step=2.0)],
                                                                                additional_constraints='Range [1.0, 13.0] Step 2.0')),
                                    ('hello_msg', 'Orcheas', ParameterDescriptor(description='The msg name to print'))
                                ])
        # parse the param values
        self.timer_interval = self.get_parameter('timer_interval').value
        self.hello_msg = self.get_parameter('hello_msg').value
        # add a callback to update on the fly
        self.add_on_set_parameters_callback(self.parameter_callback)
        self.timer = self.create_timer(timer_period_sec=self.timer_interval, callback=self.print_message)


    def parameter_callback(self, params: list[Parameter]) -> SetParametersResult:
        """Callback for setting params on the fly"""
        for param in params:
            if param.name == 'timer_interval':
                self.timer_interval = param.value
                if self.timer:
                    self.timer.cancel()
                self.timer = self.create_timer(timer_period_sec=self.timer_interval, callback=self.print_message)
            elif param.name == 'hello_msg':
                self.hello_msg = param.value
        return SetParametersResult(successful=True)


    def print_message(self):
        """Print a message based on the timer"""
        self.get_logger().info(f'Hello from {self.hello_msg} (interval={self.timer_interval})')


def main(args=None) -> None:
    """Entry Point for our node"""
    rclpy.init(args=args)   # Initialise rclpy (ROS communications)
    node = HelloParams()
    rclpy.spin(node=node)   # Keep node alive
    node.destroy_node()     # Destroy node
    rclpy.shutdown()        # Shutdown the executor


if __name__ == '__main__':
    main()

File: examples/hello_params/hello_params_cpp/src/hello_params.cpp
/**
 *  A minimal ROS2 cpp node that logs a message every second.
 *
 *  Basic structure for ROS2 node:
 *  - Initialise rclcpp
 *  - Create a node
 *  - Declare the params
 *  - Keep it alive with spin
 *  - Shutdown
 *
 *  Author: Ioannis Selinis <selinis.g@gmail.com> 2025
 */

#include <rclcpp/rclcpp.hpp>
#include <rcl_interfaces/msg/parameter_descriptor.hpp>
#include "rcl_interfaces/msg/floating_point_range.hpp"
#include <rcl_interfaces/msg/set_parameters_result.hpp>


class HelloParams : public rclcpp::Node
{
    public:
        HelloParams() : Node("hello_c_params")
        {
            rcl_interfaces::msg::ParameterDescriptor timer_interval_descriptor;
            timer_interval_descriptor.description = "Set the interval for the timer";
            timer_interval_descriptor.additional_constraints = "Range [1.0, 13.0] Step 2.0";
            timer_interval_descriptor.read_only = false;
            // automatic constraints!
            timer_interval_descriptor.floating_point_range.resize(1);
            timer_interval_descriptor.floating_point_range[0].from_value = 1.0;
            timer_interval_descriptor.floating_point_range[0].to_value   = 13.0;
            timer_interval_descriptor.floating_point_range[0].step       = 2.0;
            this->declare_parameter("timer_interval", 1.0, timer_interval_descriptor);

            rcl_interfaces::msg::ParameterDescriptor hello_msg_descriptor;
            hello_msg_descriptor.description = "The msg name to print";
            this->declare_parameter("hello_msg", "Orcheas", hello_msg_descriptor);

            m_timer_interval = this->get_parameter("timer_interval").as_double();
            m_hello_msg = this->get_parameter("hello_msg").as_string();

            m_param_cb = this->add_on_set_parameters_callback(std::bind(&HelloParams::paramCb, this, std::placeholders::_1));

            m_timer = this->create_wall_timer(std::chrono::duration<double>(m_timer_interval),
                                              std::bind(&HelloParams::printMessage, this));
            RCLCPP_INFO(this->get_logger(), "Node has started");
        }

    private:
        void printMessage()
        {
            RCLCPP_INFO(this->get_logger(), "Hello from: %s (interval=%.1f)", m_hello_msg.c_str(), m_timer_interval);
        }
        rcl_interfaces::msg::SetParametersResult paramCb(const std::vector<rclcpp::Parameter> &params)
        {
            for (const auto &param : params){
                if (param.get_name() == "timer_interval"){
                    m_timer_interval = param.as_double();
                    if (m_timer){
                        m_timer->cancel();
                    }
                    m_timer = this->create_wall_timer(std::chrono::duration<double>(m_timer_interval),
                                                      std::bind(&HelloParams::printMessage, this));
                }
                else if (param.get_name() == "hello_msg"){
                    m_hello_msg = param.as_string();
                }
            }
            rcl_interfaces::msg::SetParametersResult result;
            result.successful = true;
            result.reason = "success";
            return result;
        }
        rclcpp::TimerBase::SharedPtr m_timer;
        std::string m_hello_msg;
        double m_timer_interval;
        OnSetParametersCallbackHandle::SharedPtr m_param_cb;
};


int main (int argc, char **argv)
{
    rclcpp::init(argc, argv);                   // Initialise rclcpp (ROS communications)
    auto node = std::make_shared<HelloParams>();// Smart/Shared pointer
    rclcpp::spin(node);                         // Keep node alive
    rclcpp::shutdown();                         // Shutdown executor
}
